benchopt/utils/safe_import.py

```diff
@@ -1,31 +1,87 @@
 import sys
 import warnings
 import importlib
+import importlib.util
+
+from importlib.abc import Loader, MetaPathFinder
 from pathlib import Path
+from unittest.mock import Mock
 
 from joblib.externals import cloudpickle
 
 from ..config import RAISE_INSTALL_ERROR
 
-SKIP_IMPORT = False
+
+MOCK_ALL_IMPORT = False
 BENCHMARK_DIR = None
 PACKAGE_NAME = "benchmark_utils"
 
 
-class SkipWithBlock(Exception):
-    pass
+class MockLoader(Loader):
+    def __init__(self, name):
+        self.name = name
 
+    def create_module(self, spec):
+        return Mock(name=self.name)
 
-def skip_import():
-    """Once called, all the safe_import_context is skipped."""
-    global SKIP_IMPORT
-    SKIP_IMPORT = True
+    def exec_module(self, module):
+        module.__path__ = []
 
 
-def _unskip_import():
+class MockFinder(MetaPathFinder):
+    def __init__(self):
+        self.specs = list()
+        self.has_failed_import = False
+
+    def find_spec(self, fullname, path, target=None):
+        """
+        MockFinder is inserted at the first place of the finders list,
+        so it is the first finder called by Python.
+        The aim is to check if we can import the module {fullname},
+        if not, we mock it. If we try to import the module {fullname}
+        to check ImportError or ModuleNotFoundError,
+        the mock finder will be called again that's why we
+        store {fullname} in {self.specs} list
+        to skip the mock finder and check if the other finders throw errors.
+
+        Parameters
+        ----------
+        fullname
+        path
+        target
+
+        Returns
+        -------
+        None or a spec if the module is mocked
+        """
+        if fullname in self.specs:
+            return None
+
+        try:
+            if MOCK_ALL_IMPORT:
+                raise ImportError
+
+            self.specs.append(fullname)
+            # Check if we can import the module {fullname}
+            importlib.import_module(fullname)
+            # If the module can be imported, we let other finders to import it,
```

Comment:
I am not sure mocking only missing modules should be the goal of `mock_import`. `skip_import` skipped every import such that it was fast for benchopt to read the dependency list. Some dependencies are very large and slow to import (for example `torch` takes 1-2s to import).